### PR TITLE
free memory with mmap(PROT_NONE, MAP_FIXED)

### PR DESCRIPTION
**Pull request by vnikolai**
For platforms that does not honor madvise and it has no effect or unclear effect propose alternative solution with mmap(PROT_NONE, MAP_FIXED) to keep pages mapped but not backed up by physical memory pages. Depending on underlying behavior it might have better effect.

### CODE REVIEW

File: src/system-alloc.cc [hunk 1]
```diff
@@ -545,11 +554,19 @@ bool TCMalloc_SystemRelease(void* start, size_t length) {
...
   return false;
 }
 
 void TCMalloc_SystemCommit(void* start, size_t length) {
+#if defined(FREE_MMAP_PROT_NONE) && defined(HAVE_MMAP)
+  // remaping as MAP_FIXED to same address assuming span size did not change 
+  // since last TCMalloc_SystemRelease
+  mmap(start, length, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_FIXED,
+       -1, 0);
+#elif defined(MADV_FREE)
```
Comment (alk) [inline src/system-alloc.cc:567]: one litttle nit-pick here, why bother with elif if you can simply do else ?

**Comment by alk** [comment]
This is really nice. One little ask about replacing (in 2 place) elsif with MADV_FREE to just else which would seem more logical. Also, since madvise code handles EINTR (but yes AFAIK linux doesn't return it for mm syscalls), so might be worth doing same just for principle of least surprise.

Also thanks for making it optional. Indeed, your proposal works, but on linux both "release" and "commit" calls will take mmap_sem lock for write, while madvise (both dontneed and free) are doing it for read, and page faults which occur after empty "commit" calls also do read. So while it will fix some people's use-cases, unfortunately it cannot be default.

File: src/system-alloc.cc (hunk 1 above)
Comment (vnikolai) [reply src/system-alloc.cc]: Done

**Comment by vnikolai** [comment]
@alk did you mean handle EAGAIN same manner as madvise? EINTR confuses me here

**Comment by vnikolai** [comment]
@alk Kindly reminder please review updated version.

**Comment by alk** [comment]
Applied. Thanks. Can you clarify what OSes this is intended for ? (it works slightly less great on Linux locking-wise and  I suspect many other unix-like OSes would have same outcomes)

**Comment by vnikolai** [comment]
That is designed for QNX. And tested also by confirming memory consumption with hogs utility.

**Pull request closed without merging**